_posts/2025-12-12-pqc-on-cheriot.markdown

```diff
@@ -0,0 +1,162 @@
+Since NIST standardised a few such algorithms, there's been a lot more interest in seeing them in production, so I spent some time getting the implementations from the Linux Foundation's PQ Code Package to run on CHERIoT.
+A lot of companies are building hardware to accelerate these operations, so it seemed useful to have a performance baseline on the CHERIoT Ibex, as well as something that can be used in future CHERIoT-based products.
+
+What are ML-KEM and ML-DSA for?
+-------------------------------
+
+I am not a mathematician and so I'm not going to try to explain how these algorithms work, but I am going to explain what they're *for*.
+
+Module-Lattice-Based Key-Encapsulation Mechanism (ML-KEM) is, as the name suggests, an algorithm for key encapsulation.
+One side holds a public key and uses it (plus some entropy source) to generate a secret in both plain and encapsulated forms.
+The encapsulated secret can be sent to a remote party who holds the corresponding private key.
+The receiver can then recover unencrypted version of the secret (and detect tampering).
+Now, both parties have the same secret and can use it with some key-derivation function to produce something like an AES key for future communication.
+
+Note that this is somewhat more restrictive than traditional key-exchange protocols.
+You don't get to exchange an arbitrary value, the generation step is part of encapsulation.
+This also means that it's a fixed size, defined by the algorithm, which is why you typically feed it into a key-derivation function rather than using it directly.
+
+Module-Lattice Digital Signature Algorithm (ML-DSA) has a similarly informative name.
+It is intended for providing and validating digital signatures.
+It takes a private key, an arbitrary-sized document and context, and produces a signature.
+A holder of the associated public key can then validate that the document matches the version signed with the private key and context.
+
+These are both quite low-level building blocks for higher-level protocols.
+For example, TLS can use ML-KEM for key exchange and ML-DSA for certificate validation, but also incorporates traditional algorithms in case the PQC algorithms have unexpected weaknesses against classical computers.
+
+Initial porting
+---------------
+
+As is usually the case for CHERIoT, porting the C implementations of ML-KEM and ML-DSA required no code changes.
+I worked with upstream to slightly simplify the platform-integration layer, so we just provide a single header describing the port.
+For example, the [port heaer for ML-DSA](https://github.com/CHERIoT-Platform/cheriot-pqc/blob/main/include/mldsa_native_config.h) configures the build to produce ML-DSA44 support, defines a custom function for zeroing memory and getting entropy, and adds the `__cheriot_libcall` attribute to the all exported APIs (so we can build them as shared libraries, rather than embedded in a single compartment).
```

Comment:
```suggestion
For example, the [port here for ML-DSA](https://github.com/CHERIoT-Platform/cheriot-pqc/blob/main/include/mldsa_native_config.h) configures the build to produce ML-DSA44 support, defines a custom function for zeroing memory and getting entropy, and adds the `__cheriot_libcall` attribute to the all exported APIs (so we can build them as shared libraries, rather than embedded in a single compartment).
```